src/main/java/org/opensearch/security/dlic/rest/api/RolesApiAction.java

```diff
@@ -70,27 +72,29 @@ public ValidationResult<JsonNode> validate(RestRequest request, JsonNode jsonCon
         }
 
         private ValidationResult<JsonNode> validateMaskedFields(final JsonNode content) {
-            final ReadContext ctx = JsonPath.parse(content.toString());
-            final List<String> maskedFields = ctx.read("$..masked_fields[*]");
-            if (maskedFields != null) {
-                for (String mf : maskedFields) {
-                    if (!validateMaskedFieldSyntax(mf)) {
-                        this.validationError = ValidationError.WRONG_DATATYPE;
-                        return ValidationResult.error(RestStatus.BAD_REQUEST, this);
-                    }
-                }
+            StreamSupport.stream(content.withArray(JsonPointer.compile("/index_permissions")).spliterator(), false)
+                .flatMap(
+                    indexPermissionsNode -> StreamSupport.stream(indexPermissionsNode.withArray("/masked_fields").spliterator(), false)
+                )
+                .map(this::validateMaskedFieldSyntax)
+                .filter(Objects::nonNull)
+                .forEach(wrongMaskedField -> {
+                    this.validationError = ValidationError.WRONG_DATATYPE;
+                    wrongDataTypes.put("Masked field not valid: " + wrongMaskedField.getLeft(), wrongMaskedField.getRight());
+                });
+            if (validationError != ValidationError.NONE) {
+                return ValidationResult.error(RestStatus.BAD_REQUEST, this);
             }
             return ValidationResult.success(content);
         }
 
-        private boolean validateMaskedFieldSyntax(String mf) {
+        private Pair<String, String> validateMaskedFieldSyntax(final JsonNode maskedFieldNode) {
             try {
-                new MaskedField(mf, SALT).isValid();
+                new MaskedField(maskedFieldNode.asText(), SALT).isValid();
```

Comment:
Which of these methods throws an exception? It seems out that this calls a function called `isValid()` and expects it to throw an error if its invalid.

Reply:
Thanks for checking. I'm not ask familiar with the `MaskedField` class and can't find any good tests around it. 

This looks fine to me, but it took me a second to understand the change with returning a Pair or null vs prior to this change it returned true or false.

Reply:
Ahhh sorry for that, but the previous solution checks masked fields only ones and after that you need to repeat an attempt. So in case you have more than one invalid masked fields you have to submit PUT/PATCH more than ones. Which is not nice since in other places we try to validate (e.g. `validateDataType`) all values in the request body.

Reply:
Oops looks like when I applied a patch I missed tests which I added. I was sure that tests are part of PR :-D. Which I mentioned in the description.

Reply:
@peternied added missed tests

Reply:
@cwperks I forget to add tests when I applied the patch. I think now it should be better to understand MaskedField behave.